keps/sig-network/0006-20180314-coredns-GA-proposal.md

```diff
@@ -0,0 +1,132 @@
+
+## Summary
+
+CoreDNS is sister CNCF project and is the successor to SkyDNS, on which kube-dns is based. It is a flexible, extensible
+authoritative DNS server and directly integrates with the Kubernetes API. It can serve as cluster DNS,
+complying with the [dns spec](https://git.k8s.io/dns/docs/specification.md). As an independent project,
+it is more actively developed than kube-dns and offers performance and functionality beyond what kube-dns has. For more details, see the [introductory presentation](https://docs.google.com/presentation/d/1v6Coq1JRlqZ8rQ6bv0Tg0usSictmnN9U80g8WKxiOjQ/edit#slide=id.g249092e088_0_181), or [coredns.io](https://coredns.io), or the [CNCF webinar](https://youtu.be/dz9S7R8r5gw).
+
+Currently, we are following the road-map defined [here](https://github.com/kubernetes/features/issues/427). CoreDNS is Beta in Kubernetes v1.10, which can be installed as an alternate to kube-dns.
+The purpose of this proposal is to graduate CoreDNS to GA and be the default DNS service, replacing the current implementation of kube-dns.
+
+## Motivation
+
+* CoreDNS is more flexible and extensible than what kube-dns. 
+* CoreDNS is easily extensible and maintainable using a plugin architecture.
+* CoreDNS has fewer moving parts than kube-dns, taking advantage of the plugin architecture, making it a single executable and single process.
+* It is written in Go, making it memory-safe (kube-dns includes dnsmasq which is not). 
+* CoreDNS has [better performance](https://github.com/kubernetes/community/pull/1100#issuecomment-337747482) than [kube-dns](https://github.com/kubernetes/community/pull/1100#issuecomment-338329100) in terms of greater QPS, lower latency, and lower memory consumption. 
+* It supports a number of [use cases](https://coredns.io/2017/05/08/custom-dns-entries-for-kubernetes/) that kube-dns doesn't. As a general-purpose authoritative DNS server it has a lot of functionality that kube-dns couldn't reasonably be expected to add.
+
+### Goals
+
+* Have CoreDNS run as the default DNS server both after a new installation and an upgrade of the existing installation.
+* Make CoreDNS available as an [image in gcr.io](https://github.com/kubernetes/kubernetes/pull/59753) and define a workflow/process to update the CoreDNS versions in the future.
+* Provide a kube-dns to CoreDNS upgrade path with configuration translation.
+* Retain the configuration of a CoreDNS deployment from Alpha (1.9) or Beta (1.10).
+
+### Non-Goals
+
+* Translation of CoreDNS ConfigMap back to kube-dns (i.e., downgrade).
+* Retaining the ConfigMap of a custom/modified image of CoreDNS during upgrade. That is, if the user has modified the Corefile in the ConfigMap, those changes will not be retained (the user should back it up and redo the changes on the new ConfigMap).
+* Translation configuration of kube-dns to equivalent CoreDNS that is defined outside of the kube-dns ConfigMap. For example, modifications to the manifest or `dnsmasq` configuration.
```

Comment:
Could be considered a breaking change and needs to be documented imo.

Reply:
This should be required as there a users of kube-dns with the configmap, and their setups will break if this is not provided.

Reply:
There isn’t anything that prevents them from continuing to use Kube-dns, which is backwards compatible.

I don’t think we should claim support for arbitrary customizations, and “customizer beware” is our default stance.

Reply:
@bowei : we are converting the kube-dns ConfigMap to a CoreDns ConfigMap. this "non-goal" statement is to warn that a customized deployment of kube-dns with extra options on the cmd-lines of dnsmasq or kube-dns containers will not be translated.

in other words we support translation "StubDomains", "UpstreamNameServer" and "Federation" is supported.

Reply:
To be clear - I read this as "we WILL support graceful upgrades from the kube-dns configmap, but we will NOT support upgrade from arbitrary other config" - is that right?

Reply:
@thockin, yes. The options that can be configured in the kube-dns ConfigMap will be retained in the conversion to the Corefile. But we aren't going to tweak things for arbitrary config changes to the dnsmasq CLI options, etc. If the user has that, they should probably retain the kube-dns installation until they can upgrade in a more considered manner.